TicketingContract.sol

```diff
@@ -101,59 +87,99 @@ contract TicketPro
+
+    function loadNewTickets(uint256[] tickets) public
     {
-        for(uint i = 0; i < tickets.length; i++) 
+        require(msg.sender == organiser);
+        for(uint i = 0; i < tickets.length; i++)
         {
-            inventory[organiser].push(tickets[i]);    
+            inventory[organiser].push(tickets[i]);
         }
     }
-    
+
+    //for new tickets to be created and given over
     function passTo(uint256 expiry,
-                    uint16[] ticketIndices,
+                    uint256[] tickets,
+                    uint8 v,
+                    bytes32 r,
+                    bytes32 s,
+                    address recipient) public
+    {
+        require(msg.sender == paymaster);
+        require(expiry > block.timestamp || expiry == 0);
+        bytes32 message = encodeMessage(0, expiry, tickets);
+        address giver = ecrecover(message, v, r, s);
+        //only the organiser can authorise this
+        require(giver == organiser);
+        for(uint i = 0; i < tickets.length; i++)
+        {
+            inventory[recipient].push(tickets[i]);
+        }
+        emit PassTo(tickets, v, r, s, recipient);
+    }
+
+    //when a user sends a ticket to another but the gas is covered by the payment server
+    function passToOnBehalfOfUser(uint256 expiry,
+                    uint256[] tickets,
                     uint8 v,
                     bytes32 r,
                     bytes32 s,
-                    address recipient) public payMasterOnly
+                    address recipient) public
     {
+        require(msg.sender == paymaster);
         require(expiry > block.timestamp || expiry == 0);
-        bytes32 message = encodeMessage(0, expiry, ticketIndices);
+        bytes32 message = encodeMessage(0, expiry, tickets);
         address giver = ecrecover(message, v, r, s);
-        for(uint i = 0; i < ticketIndices.length; i++)
+        //increment givers tickets to make sure they have them
+        for(uint i = 0; i < tickets.length; i++)
         {
-            uint16 index = ticketIndices[i];
-            //needs to use revert as all changes should be reversed
-            //if the user doesnt't hold all the tickets 
-            assert(inventory[giver][index] != bytes32(0));
-            bytes32 ticket = inventory[giver][index];
-            inventory[recipient].push(ticket);
-            delete inventory[giver][index];
+            for(uint j = 0; j < tickets.length; i++)
+            {
+                bool found = false;
+                if(inventory[giver][j] == tickets[i])
+                {
+                    found = true;
+                    break;
+                }
+            }
+            //if any ticket up for sale is not found, revert state
+            //assert will revert all changes back to pre transaction
+            assert(found);
+            inventory[recipient].push(tickets[i]);
+            delete inventory[giver][i];
```

Comment:
This might be wrong. Maybe you have to make a loop to see which ticket ID in inventory[giver] == tickets[i] and delete that (i.e. the value of j above when found==true)?: delete inventory[giver][i];.

If assert() reverts delete(s), you can move the assert + push + delete into the inner for-loop

Reply:
I see your point but assert has to be outside the loop as it needs to be called if any of the tokens are not found (and thus cannot be caught inside the inner loop). I have moved the push and delete inside but the assert waits til the inner loop is finished and if any ticket is not found it will assert and revert all the changes done by the transaction.